servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyPipelinedConnection.java

```diff
@@ -0,0 +1,362 @@
+    public Single<Throwable> transportError() {
+        return connection.transportError();
+    }
+
+    @Override
+    public Completable onClosing() {
+        return connection.onClosing();
+    }
+
+    @Override
+    public Completable onClose() {
+        return connection.onClose();
+    }
+
+    @Override
+    public Completable closeAsync() {
+        return connection.closeAsync();
+    }
+
+    @Override
+    public Completable closeAsyncGracefully() {
+        return connection.closeAsyncGracefully();
+    }
+
+    @Override
+    public Channel nettyChannel() {
+        return connection.nettyChannel();
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + '(' + connection + ')';
+    }
+
+    @Override
+    public Cancellable updateFlushStrategy(final NettyConnectionContext.FlushStrategyProvider strategyProvider) {
+        return connection.updateFlushStrategy(strategyProvider);
+    }
+
+    @Override
+    public FlushStrategy defaultFlushStrategy() {
+        return connection.defaultFlushStrategy();
+    }
+
+    private void closeConnection(final Subscriber<? super Resp> subscriber, final Throwable cause) {
+        toSource(connection.closeAsync().concat(Publisher.<Resp>failed(cause))).subscribe(subscriber);
+    }
+
+    private void tryStartRead(@Nullable Subscriber<? super Resp> subscriber) {
+        if (subscriber == null) {
+            return;
+        }
+        final PublisherSource<Resp> src;
+        try {
+            src = toSource(connection.read().afterFinally(() ->
+                    tryStartRead(pollWithLockAcquired(readQueue, readQueueLockUpdater))));
+        } catch (Throwable cause) {
+            handleReadSetupError(subscriber, cause);
+            return;
+        }
+        src.subscribe(subscriber);
+    }
+
+    private final class WriteTask {
+        private final Subscriber<? super Resp> subscriber;
+        private final Publisher<Req> requestPublisher;
+        private final Supplier<FlushStrategy> flushStrategySupplier;
+        private final Supplier<WriteDemandEstimator> writeDemandEstimatorSupplier;
+
+        private WriteTask(final Subscriber<? super Resp> subscriber,
+                          final Publisher<Req> requestPublisher,
+                          final Supplier<FlushStrategy> flushStrategySupplier,
+                          final Supplier<WriteDemandEstimator> writeDemandEstimatorSupplier) {
+            this.subscriber = subscriber;
+            this.requestPublisher = requestPublisher;
+            this.flushStrategySupplier = flushStrategySupplier;
+            this.writeDemandEstimatorSupplier = writeDemandEstimatorSupplier;
+        }
+
+        void run() {
+            final PublisherSource<Resp> src;
+            try {
+                src = toSource(connection.write(requestPublisher, flushStrategySupplier,
+                        writeDemandEstimatorSupplier)
+                        .afterFinally(() -> {
```

Comment:
Does it have to be `afterFinally()` or we can use `beforeFinally()`? A positive of using `before*()` here is that any exception will bubble up to the subscriber.

Reply:
`beforeFinally` can trigger before write stream subscriber is cleaned up in some error situations so I'll just stick with `after*` for now.

Reply:
to provide more context ....
-  in the event the write publisher terminates with an error `WriteStreamSubscriber.AllWritePromise#terminateSubscriber` closes the closeChannelOutbound before propagating exception to the subscriber which results in `A write is already active on this connection.` exception on `DefaultNettyConnection`. I think the sequencing in incorrect and `closeChannelOutbound` should be after the subscriber is terminated and I will fix this. however then `merge` triggers a cancellation which results in starting the next read task, which subscribes to the `channel.read() ` publisher before the previous subscriber is "detached" from `NettyChannelPublisher` which results in a `DuplicateSubscribeException` ... in this case we want an operator like "do subscriber before, but subscription after" to invoke `tryStartRead`.
- it is also a questioning of sequencing ... if we dequeue the next element before notifying the status of the current element, and there is a synchronous type of failure/completion we will notify completion in reverse order as the queue. we have a queue so "cleaning up local state first" may result in out of order notification